File: javascript/day3/day3.js
const fs = require('fs');
const path = require('path');

const input = fs.readFileSync(path.resolve(__dirname, 'input.txt'), 'utf8');

const lines = input.split('\n');

const numbers = [];
const symbols = [];

lines.forEach((line, row) => {
  let col = 0;
  while (col < line.length) {
    const number = line.slice(col).match(/\d+/);
    if (number) {
      // console.log(number);
      numbers.push({
        number: parseInt(number[0]),
        row,
        col: col + number.index,
        colEnd: col + number.index + number[0].length,
      });
      col = col + number.index + number[0].length;
    } else {
      col = line.length;
    }
  }

  col = 0;
  while (col < line.length) {
    const symbol = line.slice(col).match(/[^0-9.]/);
    if (symbol) {
      // console.log(symbol);
      symbols.push({
        symbol: symbol[0],
        row,
        col: col + symbol.index,
      });
      col = col + symbol.index + symbol[0].length;
    } else {
      col = line.length;
    }
  }
});

function isNumberAdjacentToSymbol(row, col, colEnd) {
  const symbol = symbols.find((s) => {
    const minCol = s.col - 1;
    const maxCol = s.col + 1;
    const minRow = s.row - 1;
    const maxRow = s.row + 1;
    return (row >= minRow && row <= maxRow && ((col >= minCol && col <= maxCol) || (colEnd >= minCol && colEnd <= maxCol)));
  });
  if (symbol) {
    return true;
  }
  return false;
}

let sum = 0;
numbers.forEach((number) => {
  if (isNumberAdjacentToSymbol(number.row, number.col, number.colEnd)) {
    // console.log(number);
    sum += number.number;
  }
});

console.log(sum);

File: lua/day3/day3.lua
local open = io.open

local function read_file(path)
  local file = open(path, "rb")  -- r read mode and b binary mode
  if not file then return nil end
  local content = file:read "*a" -- *a or *all reads the whole file
  file:close()
  return content
end

local fileContent = read_file("lua/day3/input.txt");

if not fileContent then
  print("File not found")
  os.exit()
end

-- print(fileContent)

-- fileContent looks like this:
-- 467..114..
-- ...*......
-- ..35..633.
-- ......#...
-- 617*......
-- .....+.58.
-- ..592.....
-- ......755.
-- ...$.*....
-- .664.598..

-- local space = '.'
-- local symbolTypes = { '#', '*', '$', '+', '%', '@', '-', '/', '=', '&' }

local numbers = {}
local symbols = {}

local function getNextSymbol(line, index)
  local nextIndex = line:find("[^%d.]", index)
  if not nextIndex then
    return nil, nil
  end
  return nextIndex, line:sub(nextIndex, nextIndex)
end

local row = 0
for line in fileContent:gmatch("[^\r\n]+") do
  row = row + 1
  local col = 1
  for number in line:gmatch("%d+") do
    col = line:find(number, col)
    -- print(number, row, col)
    table.insert(numbers, { number = tonumber(number), row = row, col = col, endCol = col + #number - 1 })
  end

  col = 1
  while col <= #line do
    local nextIndex, symbol = getNextSymbol(line, col)
    if symbol and nextIndex <= #line then
      -- print(symbol, row, nextIndex)
      table.insert(symbols, { symbol = symbol, row = row, col = nextIndex })
      col = nextIndex + 1
    else
      col = #line + 1
    end
  end
end

local function isNumberAdjacentToSymbol(row, col, endCol)
  for _, symbol in ipairs(symbols) do
    local minSymbolRow = symbol.row - 1
    local maxSymbolRow = symbol.row + 1
    local minSymbolCol = symbol.col - 1
    local maxSymbolCol = symbol.col + 1
    if row >= minSymbolRow and row <= maxSymbolRow and
        ((col >= minSymbolCol and col <= maxSymbolCol) or (endCol >= minSymbolCol and endCol <= maxSymbolCol)) then
      return true
    end
  end
  return false
end

print(#symbols)

local sum = 0
for _, number in ipairs(numbers) do
  if isNumberAdjacentToSymbol(number.row, number.col, number.endCol) then
    sum = sum + number.number
  end
end

print(sum)
